### PR TITLE
Add oauth2 direct token authentication

### PR DESCRIPTION
**Pull request by staticfloat**
This change allows the user to submit a `token` directly to `/auth/login` to bypass the OAuth2 web login flow.  This is especially useful if buildbot is being accessed by an automated process that has already obtained a [personal access token](https://developer.github.com/v3/oauth_authorizations/) or similar.

This has been tested working with GitHub, I haven't tested with Gitlab or any other OAuth2 providers, but I don't see any reason why this implementation wouldn't be generic.

### CODE REVIEW

**Comment by staticfloat** [comment]
> There is a need for unit tests.

I've added something in the `test_GithubVerifyCode`  unit test.  Is this what you'd want?

> Also, it would be good to have a small paragraph in the doc

Where are the docs?  Are they in this repository?

File: master/buildbot/test/unit/test_www_oauth.py [hunk 1]
```diff
@@ -161,6 +161,11 @@ def test_GithubVerifyCode(self):
                           'username': 'bar',
                           'groups': ["hello", "grp"],
                           'full_name': 'foo bar'}, res)
+        res = yield self.githubAuth.acceptToken("TOK3N")
```
Comment (tardyp) [inline master/buildbot/test/unit/test_www_oauth.py]: please put this in a separate test.

File: master/buildbot/newsfragments/github_access_token.feature [hunk 2]
```diff
@@ -0,0 +1 @@
+:py:class: `~buildbot.util.OAuth2LoginResource` now supports the `token` URL parameter.  If a user wants to authenticate through OAuth2 with a pregenerated token (such as the `access_token` provided by GitHub) it can be passed to `/auth/login` as the `token` URL parameter and the user will be authenticated to buildbot with those credentials.
```
Comment (tardyp) [inline master/buildbot/newsfragments/github_access_token.feature]: pregenerated is not like by the spell checker. probably pre-generated will be better

**Comment by tardyp** [comment]
also, the docs are in master/docs

probably you will need to add a paragraph in the https://github.com/buildbot/buildbot/blob/master/master/docs/developer/rest.rst#L1 file

File: master/buildbot/test/unit/test_www_oauth.py [hunk 3]
```diff
@@ -161,6 +161,28 @@ def test_GithubVerifyCode(self):
                           'username': 'bar',
                           'groups': ["hello", "grp"],
                           'full_name': 'foo bar'}, res)
+    @defer.inlineCallbacks
```
Comment (tardyp) [inline master/buildbot/test/unit/test_www_oauth.py]: need a empty line above

File: master/docs/developer/rest.rst [hunk 4]
```diff
@@ -162,6 +162,18 @@ A simple example:
...
+~~~~~~~~~~~~~~
+
+Authentication to the REST API is performed in the same manner as authentication to the main web interface.  Once credentials have been established, a cookie will be set, which must be sent to the buildbot REST API with every request thereafter.  For those buildbot instances using OAuth2 authentication providers, access tokens can be used for automated access.  For example, GitHub's personal access tokens can be used to access the buildbot as a github user without needing to store the username and password of the user.  To use an OAuth2 access token, send a `GET` request to the `/auth/login` with the `token` URL parameter set to the access token that the OAuth2 provider has given you.  A `python` example using `requests` is shown below, where we first authenticate with our OAuth2 access token, and then are able to request otherwise shielded endpoints:
+
+.. code-block:: python
+
+    import requests
+    s = requests.Session()
+    s.get("https://<buildbot_url>/auth/login", params={"token": OAUTH_TOKEN})
+    builders = s.get("https://<buildbot_url>/api/v2/builders").json()
```
Comment (tardyp) [inline master/docs/developer/rest.rst]: pretty cool thanks!

File: master/buildbot/test/unit/test_www_oauth.py (hunk 1 above)
Comment (rodrigc) [inline master/buildbot/test/unit/test_www_oauth.py]: I have seen that StopIteration error before, but not lately.  I'm not sure what is causing it.

**Comment by tardyp** [comment]
please just run "make isort" this will make sure that the pep8 is right.
There is another empty line issue.

File: master/buildbot/test/unit/test_www_oauth.py (hunk 1 above)
Comment (rodrigc) [inline master/buildbot/test/unit/test_www_oauth.py]: I am seeing GeneratorExit errors in the Python 3 build log:
https://nine.buildbot.net/#/builders/2/builds/109615/steps/5/logs/stdio

They are not causing the build to fail, but they need to be investigated at some point.

**Comment by staticfloat** [comment]
🎉

**Pull request merged**